### ISSUE TITLE
Exclude issues from summary, if not listed in Stash Diff

### ISSUE THREAD

**Issue by kirkor** [AmadeusITGroup/sonar-stash#58]
If your branch is behind the master and on master some issues has been fixed,  those issues will be displayed in you PR in summary comment. 

The summary is about new issues on your branch, my proposition is to add boolean property to exclude from summary comment issues that are not listed in Stash Diff.

I will develop this by myself here: https://github.com/kirkor/sonar-stash/tree/issues_only_from_diff

**Comment by re-schneider** [comment]
Hi,

what you describe reminds me of issue #26 that is partially linked to PR #24.

Can you please check how much it matches your situation. This way, this issue can be more specific (if something else is not addressed yet) and you can benefit from code already done (if the PR contains changes you need).

Thanks in advance for the update :)

**Comment by kingoleg** [comment]
The issue and https://github.com/AmadeusITGroup/sonar-stash/issues/25 are in SonarQubeCollector, line 64:


```java
        InputFile inputFile = inputFileCache.getInputFile(issue.componentKey());
        if (inputFile == null){
          LOGGER.debug("Issue {} is not linked to a file, not added to the report", issue.key());
        } else {
          String path = new PathResolver().relativePath(projectBaseDir, inputFile.file());
             
          // Create the issue and Add to report
          SonarQubeIssue stashIssue = new SonarQubeIssue(key, severity, message, rule, path, line);
          result.add(stashIssue);
        } 
      }
```

Even you try to found case there there is no path with if, still after PathResolver#relativePath you can get null. I propose to add a new check

PR is on the way

**Comment by t-8ch** [comment]
@kingoleg This is fixed in the new branch I am going to push soon.
(The validation succeeded just 5 minutes ago)

**Comment by kingoleg** [comment]
I fixed this file,SonarQubeCollector, I see the "No new issues detected!" PR summary comment, but `build	21-Mar-2017 18:06:32	[INFO] Pull-request 961 (TEST/test) NOT APPROVED by user "sonarqube"`

**Comment by kingoleg** [comment]
Ok, I have an issuw with coverage. Coverage is lowered, but no issue is added to PR per file in diff or in summary comment. New bug?

```
build	22-Mar-2017 00:25:49	[INFO] ANALYSIS SUCCESSFUL
build	22-Mar-2017 00:25:49	[INFO] Executing post-job org.sonar.plugins.stash.StashIssueReportingPostJob
build	22-Mar-2017 00:25:51	[INFO] SonarQube issues reported to Stash by user "sonarqube" have been reset
build	22-Mar-2017 00:25:51	[INFO] New SonarQube issues have been reported to Stash.
build	22-Mar-2017 00:25:51	[INFO] SonarQube analysis overview has been reported to Stash.
build	22-Mar-2017 00:25:51	[INFO] TODO issueNumber 0
build	22-Mar-2017 00:25:51	[INFO] TODO issueTotal 0
build	22-Mar-2017 00:25:51	[INFO] TODO can approve true
build	22-Mar-2017 00:25:51	[INFO] TODO coverageReport.countLoweredIssues() 0
build	22-Mar-2017 00:25:51	[INFO] TODO coverageReport.getEvolution() -16.7
build	22-Mar-2017 00:25:51	[INFO] TODO decline!
build	22-Mar-2017 00:25:51	[INFO] Pull-request 979 (TEST/test) NOT APPROVED by user "sonarqube"
```

**Comment by kingoleg** [comment]
It looks like the case is next:

1. I have 16.7% coverage on the project on full sonar branch scan
2. I have a class without tests at all (gwt stuff) and I already have an issue in full sonar branch scan about it
3. I add some more lines into the class, and do not add tests
4. I make a PR with it
5. I have: 16.7% previous coverage, 0 current coverage, negative coverageReport.getEvolution()
6. PR is not approved, no issue is added to PR, no info about coverageReport.getEvolution() in summary

WIE:
If there is no coverage issue and coverageReport.getEvolution() is negatiive - we should approve
Or
Any parameter to sonar-stash to turn of coverageReport.getEvolution()
And in any case
Summary comment should said something about it

In other words: if we have a class without tests, and we do not add a test on it - PR should be approved whatever is coverage degradation.

For sure, we can disable gwt classes from sonar, but if there any case to remove only coverage for them? If yes, it is also can be a workaround.

**Comment by t-8ch** [comment]
@kingoleg can you try the code in PR #106 
This replaces the coverage functionality.

**Comment by kingoleg** [comment]
I will try to try. But look, it has a broken build according to travis

Other issue with coverage:
1. I have 16.7% coverage on the project on full sonar branch scan
2. I changed pom.xml
3. I make a PR with it
4. I have: 16.7% previous coverage, 0 current coverage, negative coverageReport.getEvolution()
5. PR is not approved, no issue is added to PR, no info about coverageReport.getEvolution() in summary

WIE
Files that cannot be cover with tests has 100% coverage
Or
Do not participate in coverage calculation

**Comment by kingoleg** [comment]
https://github.com/AmadeusITGroup/sonar-stash/pull/106 looks much better then master. I will continue to test it.

**Comment by kingoleg** [comment]
Some thoughts about #106 

1. It should be able to choose line or branch coverage evolution to check or both
2. The plugin should respect common-java:InsufficientBranchCoverage, common-java:InsufficientLineCoverage setup, minimumBranchCoverageRatio and/or minimumLineCoverageRatio respectively. It the file coverage degrades but still is more then the minimum ratio, PR should be approved. It should not be a fight for the last coverage ratio.

**Comment by kingoleg** [comment]
Ok, it pass my tests, I am looking forward for clean issues in #106 and at least minimumLineCoverageRatio or even new sonar.stash.task.coverage.threshold property

**Comment by t-8ch** [comment]
About the behaviour
`"It the file coverage degrades but still is more then the minimum ratio, PR should be approved."`

I would keep the behaviour in the current refactoring. But this proposal would definitively a candidate for a future feature. (Both are).
Can you open feature requests for those?

(And just in case this wasn't obvious: I very much appreciate your feedback @kingoleg!)

**Comment by kingoleg** [comment]
Opened #109 and #110

**Comment by sidd9433** [comment]
Is this resolved? I still have this problem in 1.3.0.

**Comment by t-8ch** [comment]
@kirkor @sidd9433 
I am not sure about hiding issues that are not in the diff view.
Issues that are not linked to the diff view may be either valid or invalid. (See explanation below).
In both cases hiding it from the overview comment will lead to inconsistencies and confused users.

A valid issue that is not shown in the diff view but the overview may be file- or project-wide issues.
Hiding them from the overview will be incorrect.

An invalid issue not shown in the diff view may result from mis- or reconfigurations and inconsistencies in the setup.
In this case the only solution is to fix the root cause. Some of those invalid issues may very well appear withing the diff view and would still be shown, also confusing users.

**Comment by t-8ch** [comment]
(The other issues described in this thread should be fixed)

**Comment by sidd9433** [comment]
@t-8ch 
In that case, may be a boolean property will be useful so that people can decide how they want to see the overview ("All new issues" or "only issues available in diff view").
Can this be done?

**Comment by t-8ch** [comment]
@sidd9433 It is of course possible.
However this will add yet another option that has to be maintained and will still produce results that are wrong and confusing.
Is it impossible for you to make sure that the builds run in the correct order?
